### DISCUSSION TITLE
Git push still linked to old Github account

### DISCUSSION THREAD

**Discussion by mjaffery96** [community/community#41882] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

I made a new github account with a different username and email. When trying to push a commit on a project, git said

-----
ERROR: Permission to new-account/project denied to old-account.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
-----

Checking git log shows this

-----
commit d830f86a4c04c0b8b620b1266b3121faf9daeea8 (HEAD -> main)
Author: name email@gmail.com
Date:   Thu Dec 15 18:46:13 2022 -0800

    Add index.html

commit 659e5fcff2cc18f5c187fa5d139a951672a14871 (origin/main, origin/HEAD)
Author: name email@users.noreply.github.com
Date:   Thu Dec 15 16:41:50 2022 -0800
-----

So then I tried changing the user.email to the email@users.noreply.github.com just in case but after trying to push again it still shows the first message. 

Is there anything else I can do to fix this?

**Comment by getlost01** [comment]
I prefer to remove the previous git repo using 
`rm -fr .git` 
and then try to reinitialize the git repo and push in a new GitHub id.


or

do a clone of the previous git repo and push in the new GitHub account.

**Reply by mjaffery96** [reply]
I'm sorry, but could you explain what you mean by pushing in a new github id/account? Sorry I'm still new to using git

**Comment by airtower-luna** [comment]
It sounds like you have credentials for your old account configured, you'll need to remove those and then provide credentials for your new account. Exactly how to do that depends on your OS, configuration, and the protocol (HTTPS or SSH) you're using to push.

> So then I tried changing the user.email to the [email@users.noreply.github.com](mailto:email@users.noreply.github.com) just in case but after trying to push again it still shows the first message.

`user.email` is used to set the author email in your commits. That's what GitHub uses to detect which account a commit belongs to, but it has nothing to do with authorizing pushes. You can push commits from any author if you have write access to the repository.

**Reply by mjaffery96** [reply]
> It sounds like you have credentials for your old account configured, you'll need to remove those and then provide credentials for your new account. Exactly how to do that depends on your OS, configuration, and the protocol (HTTPS or SSH) you're using to push.

Do you know any good resources where I could learn to do that? I've tried searching but couldn't find anything that helped. I'm working with windows if that helps

**Reply by airtower-luna** [reply]
The [documentation on Git Credential Storage](https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage) mentions that Git for Windows uses the Windows Credential Store. I don't use Windows myself, so I can't tell you how to work with that, but it's a thing to look up, or maybe someone who uses it can comment here.

**Reply by Marvischief** [reply]
> The [documentation on Git Credential Storage](https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage) mentions that Git for Windows uses the Windows Credential Store. I don't use Windows myself, so I can't tell you how to work with that, but it's a thing to look up, or maybe someone who uses it can comment here.

Please I use mac and facing the same issue, what can I do

**Reply by JAlexMcGraw** [reply]
I have followed all of these steps, and I'm still receiving an error that push permissions are being denied to my old github account, and not the one I configured it to. Could it be because I'm pushing from a .venv?

**Reply by ninditobing7** [reply]
> * git config --system --unset credential.helper

it works

**Comment by Ezio-commits** [comment]
Same error is coming even after i changed the config --global user.name and user.email

**Reply by airtower-luna** [reply]
Those settings have nothing to do with authentication, they only set the author name and email that'll be recorded in commits you make.

You need to fix the username/PAT combination or SSH key you use to connect to GitHub.

**Reply by Rishabh-Commits** [reply]
Are you talking about ```gh auth login```?

**Reply by airtower-luna** [reply]
No, I'm talking about what you enter (and possibly cache) when Git asks for your username and password e.g. on `git push` (if you're using HTTPS), or your SSH setup (if you're using SSH). The token `gh` acquires should work for Git push/fetch operations too, though.

**Comment by Rishabh-Commits** [comment]
What if it's not even asking pass

**Reply by airtower-luna** [reply]
Then your credentials are probably stored *somewhere* on your machine, see the [discussion above](https://github.com/orgs/community/discussions/41882#discussioncomment-4426003) and especially the [Git Credential Storage](https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage) documentation.

**Reply by Ezio-commits** [reply]
Yeah thanks. Fixed it! 
As a newbie it took me a while to get it :(

**Reply by connoromalley** [reply]
how did you fix it?

**Comment by RONIE70** [comment]
created a new GitHub account and did a git push --setup-upstream origin after I added my origin branch.

After deleting my old GitHub credentials under Credential Manager in Windows, it seemed to be working fine, was getting the GitHub-sign-in window from Visual Studio Code to sign in and used new account credentials.

Once I did git push to push my project upstream, my GitHub now tells me my old account had pushed the project into my new account's branch...

How is that possible? I don't have any SSH or other tokens assigned from my old account so that is the confusing part.

I also can't seem to find a solution to "log out" or get rid of my old GitHub from upstreaming this to my new account. Is there something I am missing?

**Reply by airtower-luna** [reply]
Your credentials are only for authorizing the push (allowing you to push to the repository), they have *nothing* to do with who is the author of the commits you push. Pushing commits authored by someone else is normal and expected behavior.

When you make a commit, Git records the author name (`user.name`) and email address (`user.email`) you have configured in the commit, see [Your Identity](https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup#_your_identity) in the Git setup documentation.

GitHub (and all other Git hosting platforms I'm aware of) take the author email recorded in commits and match it against email addresses in their user database, and if there's a match show that user as the author of the commit.

So: Update your `user.email` and maybe `user.name`, and your commits should show up right. :slightly_smiling_face:

**Comment by francomac** [comment]
something that has worked for me in the past is: recreating the whole process to create a new `id_rsa` file from scratch, to be sure that I'm using the right email. just in case, remember to re-add your new `rsa` file identity at the end: `ssh-add  ~/.ssh/<YOUR_GENERATED_RSA_FILE>`

I frequently use different `id_rsa` files, which is why I know about this issue. hope it helps.

**Comment by ChidiChuks** [comment]
It appears that you're encountering permission issues when pushing to a GitHub repository after changing your username and email. The error message you're seeing, "Permission to new-account/project denied to old-account," indicates that Git is still trying to use your old account credentials when interacting with the repository.

Here are steps to resolve this issue:

1. **Update Your Git Configuration:**

   Make sure you've updated both your `user.name` and `user.email` in your local Git configuration to match your new GitHub account credentials. You can set or update them using the following commands:

   ```bash
   git config --global user.name "Your New Username"
   git config --global user.email "your_new_email@example.com"
   ```

   Replace "Your New Username" and "your_new_email@example.com" with the correct values associated with your new GitHub account.

2. **Check Remote URL:**

   Verify that the remote URL for the repository is set correctly to your new account. You can check the remote URL using the following command:

   ```bash
   git remote -v
   ```

   If the remote URL still points to your old account, you can update it using the following command:

   ```bash
   git remote set-url origin https://github.com/new-account/project.git
   ```

   Replace `https://github.com/new-account/project.git` with the URL of your new GitHub repository.

3. **Generate and Use SSH Keys (Optional):**

   If you're still facing issues and you've recently changed your GitHub account, it's a good idea to use SSH keys for authentication. Follow these steps to generate and use SSH keys:

   - Generate an SSH key pair if you haven't already:

     ```bash
     ssh-keygen -t ed25519 -C "your_email@example.com"
     ```

   - Add your SSH key to the SSH agent:

     ```bash
     eval "$(ssh-agent -s)"
     ssh-add ~/.ssh/id_ed25519
     ```

   - Add your SSH public key (`~/.ssh/id_ed25519.pub`) to your new GitHub account's SSH keys in your GitHub account settings.

   - Change the remote URL to use the SSH URL:

     ```bash
     git remote set-url origin git@github.com:new-account/project.git
     ```

4. **Token Authentication (For HTTPS):**

   If you are using HTTPS for Git authentication and are still facing issues, you can try using a personal access token. Create a personal access token in your new GitHub account and use it as your password when pushing changes over HTTPS.

   - When prompted for a password, use your personal access token instead of your GitHub account password.

5. **Check Repository Permissions:**

   Ensure that your new GitHub account has the necessary permissions (read/write access) to the repository. You can check this in the repository's settings under "Manage access."

After following these steps, you should be able to push to your GitHub repository using your new account credentials without encountering the permission error. If you continue to experience issues, please provide more details about the specific error message, and I'll be happy to assist further.

**Comment by Scotonax** [comment]
I had a credential problem too. Here how I solved:
1. First I deleted the old account's credential stored on my computer to add the new one. Here a video that shows how to do that: [https://youtu.be/fkKkojKYZ8U?si=BXBNTt-muU3OD4rM](url)
2. Then I had another problem (I don't know if it will be your case too or if you will be good only by following the first step). I was able to pull and push repositories form the new account, but the pushes appeared to be made from the old account. To solve that I had to manually set the new account username and email. To do that open the command prompt form the project directory and use these commands
`
$ git config --global user.name {YOUR GITHUB ACCOUNT USERNAME}
`
`
$ git config --global user.email {TOUR GITHUB ACCOUNT EMAIL}
`
That should resolve the problem (or at least it worked for me)

**Reply by isaactobiloba** [reply]
For anyone still encountering this issue, please copy the YouTube link posted. That's all you need.

https://youtu.be/fkKkojKYZ8U?si=BXBNTt-muU3OD4rM

**Comment by ChewyToast** [comment]
I don't know if you use HTTPS, SSH, or GitHub CLI, and I don't know the exact working of HTTPS and GitHub CLI, but in the case of SSH it's not enough that the author of the commit is the same username, you need to have configured the ssh key in the keys section of your profile. You must first generate a key with the command:
```ssh-keygen```.
Then copy the contents of the id_rsa.pub file and add the key in the settings. This way you will have permission to push to the repository.

**Reply by leeringrose** [reply]
@mjaffery96 
I recommend using Github Desktop to fix this problem.
You can find `Options...` menu by clicking the `File` from the top menu bar.
![image](https://github.com/community/community/assets/132311826/97252fef-44ff-45a6-bb45-7b63fe7533fd)
